lib/src/idx/trees/mtree.rs

```diff
@@ -284,53 +303,108 @@ impl MTree {
 		&mut self,
 		tx: &mut Transaction,
 		store: &mut MTreeNodeStore,
-		v: Vec<Number>,
+		obj: Vec<Number>,
 		id: DocId,
 	) -> Result<(), Error> {
+		#[cfg(debug_assertions)]
+		debug!("Insert - obj: {:?} - doc: {}", obj, id);
+		let obj = Arc::new(obj);
+		// First we check if we already have the object. In this case we just append the doc.
+		if self.append(tx, store, obj.clone(), id).await? {
+			return Ok(());
+		}
 		if let Some(root_id) = self.state.root {
 			let node = store.get_node(tx, root_id).await?;
-			if let InsertionResult::PromotedEntries(r1, r2) =
-				self.insert_at_node(tx, store, node, &None, Arc::new(v), id).await?
+			// Otherwise, we insert the object with possibly mutating the tree
+			if let InsertionResult::PromotedEntries(o1, p1, o2, p2) =
+				self.insert_at_node(tx, store, node, &None, obj, id).await?
 			{
-				self.create_new_internal_root(store, r1, r2)?;
+				self.create_new_internal_root(store, o1, p1, o2, p2)?;
 			}
 		} else {
-			self.create_new_leaf_root(store, v, id)?;
+			self.create_new_leaf_root(store, obj, id)?;
 		}
 		Ok(())
 	}
 
 	fn create_new_leaf_root(
 		&mut self,
 		store: &mut MTreeNodeStore,
-		v: Vec<Number>,
+		obj: Arc<Vec<Number>>,
 		id: DocId,
 	) -> Result<(), Error> {
 		let new_root_id = self.new_node_id();
 		let p = ObjectProperties::new_root(id);
-		let mut objects = LeafIndexMap::with_capacity(1);
-		objects.insert(Arc::new(v), p);
+		let mut objects = LeafMap::new();
+		objects.insert(obj, p);
 		let new_root_node = store.new_node(new_root_id, MTreeNode::Leaf(objects))?;
 		store.set_node(new_root_node, true)?;
-		self.state.root = Some(new_root_id);
-		self.updated = true;
+		self.set_root(Some(new_root_id));
 		Ok(())
 	}
 
 	fn create_new_internal_root(
 		&mut self,
 		store: &mut MTreeNodeStore,
-		r1: RoutingEntry,
-		r2: RoutingEntry,
+		o1: Arc<Vector>,
+		p1: RoutingProperties,
+		o2: Arc<Vector>,
+		p2: RoutingProperties,
 	) -> Result<(), Error> {
 		let new_root_id = self.new_node_id();
-		let new_root_node = store.new_node(new_root_id, MTreeNode::Internal(vec![r1, r2]))?;
+		#[cfg(debug_assertions)]
+		debug!(
+			"New internal root - node: {} - r1: {:?}/{} - r2: {:?}/{}",
+			new_root_id,
+			o1.as_ref(),
+			p1.radius,
+			o2.as_ref(),
+			p2.radius
+		);
+		let mut entries = InternalMap::new();
+		entries.insert(o1, p1);
+		entries.insert(o2, p2);
+		let new_root_node = store.new_node(new_root_id, MTreeNode::Internal(entries))?;
 		store.set_node(new_root_node, true)?;
-		self.state.root = Some(new_root_id);
-		self.updated = true;
+		self.set_root(Some(new_root_id));
 		Ok(())
 	}
 
+	async fn append(
+		&self,
+		tx: &mut Transaction,
+		store: &mut MTreeNodeStore,
+		object: Arc<Vector>,
```

Comment:
Nit: I'm a bit confused on potential term difference between the paper and the implementation!
So- what's in the node? Objects, or entries? Seems like the paper uses "entries" and the implementation uses "objects"?

Reply:
And if we use a term that is different from the paper, I'd guess it would be awesome to mention that difference in code comment!

Something like `we use objects instead of entries in this implementation- a node contains "entries" in the paper. it contains "objects" in this impl` or something like that. (I believe you have a better explanation than mine!

Reply:
@emmanuel-keller That makes sense! Thank you very much for the awesome explanation.

I wish whole your last comment turned into code comment, so that we don't need to ask/guess/etc about that!

Reply:
I've added comment in the code. I also have aliased the type Arc<Vector> with the name SharedVector to improve the readability